src/UnitTests/Core/Impl/Threading/UIThreadHelper.cs

```diff
@@ -221,19 +258,95 @@ public class UIThreadHelper {
+                return new CallSafeResult<T> {Exception = ExceptionDispatchInfo.Capture(tae)};
             } catch (Exception e) {
-                return new CallSafeResult<T> { Exception = ExceptionDispatchInfo.Capture(e) };
+                return new CallSafeResult<T> {Exception = ExceptionDispatchInfo.Capture(e)};
             }
         }
 
         private class CallSafeResult<T> {
             public T Value { get; set; }
             public ExceptionDispatchInfo Exception { get; set; }
         }
+
+
+        int IMainThread.ThreadId => Thread.ManagedThreadId;
+
+        void IMainThread.Post(Action action, CancellationToken cancellationToken) {
+            var bl = _blockingLoop.Value;
+            if (bl != null) {
+                bl.Post(action);
+            } else {
+                InvokeAsync(action, cancellationToken).DoNotWait();
+            }
+        }
+
+        private class BlockingLoop {
+            private readonly Func<Task> _func;
+            private readonly AutoResetEvent _are;
+            private readonly ConcurrentQueue<Action> _actions;
+
+            public Task Task { get; private set; }
+
+            public BlockingLoop(Func<Task> func) {
+                _func = func;
+                _are = new AutoResetEvent(false);
+                _actions = new ConcurrentQueue<Action>();
+            }
+
+            public void Start() {
+                Task = _func();
+                Task.ContinueWith(Complete);
+                while (!Task.IsCompleted) {
+                    _are.WaitOne();
+                    ProcessQueue();
+                }
+            }
+
+            public void ProcessQueue() {
+                Action action;
+                while (_actions.TryDequeue(out action)) {
+                    action();
+                }
+            }
+
+            private void Complete(Task task) {
+                _are.Set();
+            }
+
+            // TODO: Add support for cancellation token
+            public void Post(Action action) {
+                _actions.Enqueue(action);
+                _are.Set();
+            }
+        }
+
+        private class BlockingLoopSynchronizationContext : SynchronizationContext {
+            private readonly UIThreadHelper _threadHelper;
+
+            public BlockingLoopSynchronizationContext(UIThreadHelper threadHelper) {
+                _threadHelper = threadHelper;
+            }
+
+            public override void Send(SendOrPostCallback d, object state) {
+                _threadHelper._application.Dispatcher.Invoke(DispatcherPriority.Send, d, state);
+            }
+
+            public override void Post(SendOrPostCallback d, object state) {
+                var bl = _threadHelper._blockingLoop.Value;
+                if (bl != null) {
+                    bl.Post(() => d(state));
+                } else {
+                    _threadHelper._application.Dispatcher.BeginInvoke(DispatcherPriority.Send, d, state);
```

Comment:
We don't load VS threading assemblies in tests cause they will make tests much slower and `ThreadHelper` don't provide enough level of control. This code emulates `JoinableTaskFactory` behavior - it creates a blocking loop that can be accessed only from the code that has specific `AsyncLocal` value. `JoinableTaskFactory` does pretty much the same with much bigger amount of code and with respect to COM messages.